### ISSUE TITLE
Unreliable slow subscriber skipping too much data

### ISSUE THREAD

**Issue by kostya-sh** [real-logic/aeron#541]
Currently an image goes unavailable and becomes available again for subscriptions that cannot keep up with the publisher in multi subscription environment. When this happens the subscriber is not getting any data for the `aeron.image.liveness.timeout` nanoseconds (10 seconds by default). This seems to be overkill for an unreliable (reliable=false) subscription where the gap can be filled immediately after receiving flow control over run.

So if I run 2 RateSubscribers from aeron-samples:
```
java \
   -Daeron.dir=C:\tmp\RateSubscriber-1 \
   -Daeron.sample.embeddedMediaDriver=true \
   -Daeron.sample.channel=aeron:udp?control=localhost:5555|reliable=false \
  io.aeron.samples.RateSubscriber

java \
   -Daeron.dir=C:\tmp\RateSubscriber-2 \
   -Daeron.sample.embeddedMediaDriver=true \
   -Daeron.sample.channel=aeron:udp?control=localhost:5555|reliable=false \
  io.aeron.samples.RateSubscriber
```

And then run StreamingPublisher:
```
java \
   -Daeron.dir=C:\tmp\StreamingPublisher \
   -Daeron.sample.embeddedMediaDriver=true \
   -Daeron.sample.channel=aeron:udp?control=localhost:5555 \
  io.aeron.samples.StreamingPublisher
```

I can see that the faster subscriber gets all the messages:
```
0.0 msgs/sec, 0.0 payload bytes/sec, totals 0 messages 0 MB
Available image on aeron:udp?control=localhost:5555|reliable=false streamId=1 sessionId=722287384 from 127.0.0.1:5555
1.3e+05 msgs/sec, 3.4e+07 payload bytes/sec, totals 133261 messages 32 MB
1.9e+05 msgs/sec, 4.9e+07 payload bytes/sec, totals 324374 messages 79 MB
2.3e+05 msgs/sec, 5.9e+07 payload bytes/sec, totals 555832 messages 135 MB
2.3e+05 msgs/sec, 6.0e+07 payload bytes/sec, totals 790161 messages 192 MB
2.0e+05 msgs/sec, 5.1e+07 payload bytes/sec, totals 987649 messages 241 MB
1.2e+04 msgs/sec, 3.2e+06 payload bytes/sec, totals 1000000 messages 244 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 1000000 messages 244 MB
```

And the slower subscriber looses almost all messages:
```
0.0 msgs/sec, 0.0 payload bytes/sec, totals 0 messages 0 MB
Available image on aeron:udp?control=localhost:5555|reliable=false streamId=1 sessionId=722287384 from 127.0.0.1:5555
3.1e+02 msgs/sec, 7.9e+04 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
Unavailable image on aeron:udp?control=localhost:5555|reliable=false streamId=1 sessionId=722287384
Available image on aeron:udp?control=localhost:5555|reliable=false streamId=1 sessionId=722287384 from 127.0.0.1:5555
0.0 msgs/sec, 0.0 payload bytes/sec, totals 307 messages 0 MB
```

I think this happens because once it gets the first flow control over run there is no way it can catch up and all subsequent messages overrun as well. As result it has to wait for 10 seconds for image timeout to occur loosing all messages during this time.

AeronStat from the slow subscriber shows large number of over runs:
```
13:44:43 - Aeron Stat (CnC v13), pid 21784
======================================================================
  0:                    0 - Bytes sent
  1:          262,917,760 - Bytes received
  2:                    0 - Failed offers to ReceiverProxy
  3:                    0 - Failed offers to SenderProxy
  4:                    0 - Failed offers to DriverConductorProxy
  5:                    0 - NAKs sent
  6:                    0 - NAKs received
  7:                  637 - Status Messages sent
  8:                    0 - Status Messages received
  9:                    0 - Heartbeats sent
 10:                1,004 - Heartbeats received
 11:                    0 - Retransmits sent
 12:                    0 - Flow control under runs
 13:              228,443 - Flow control over runs
 14:                    0 - Invalid packets
 15:                    0 - Errors
 16:                    0 - Short sends
 17:                    0 - Sender flow control limits applied
 18:                    0 - Unblocked Publications
 19:                    0 - Unblocked Control Commands
 20:                    0 - Possible TTL Asymmetry
 21:                    0 - ControllableIdleStrategy status
 22:                    1 - Loss gap fills
 23:                    1 - rcv-channel: aeron:udp?control=localhost:5555|reliable=false
 24:    1,533,822,282,867 - client-heartbeat: 0
 28:          288,001,088 - sub-pos: 1 722287384 1 aeron:udp?control=localhost:5555|reliable=false @288001088
 29:          288,001,088 - rcv-hwm: 3 722287384 1 aeron:udp?control=localhost:5555|reliable=false
 30:          288,001,088 - rcv-pos: 3 722287384 1 aeron:udp?control=localhost:5555|reliable=false
```

Would it make more sense for an unreliable subscriber to gap fill immediately on flow control overrun? This way it will loose less messages.

The same situation can of course happen with reliable subscriptions as well. If I understand how aeron works correctly a reliable subscription can recover from the overrun by sending NACK to the publisher given that term buffer is big enough. I would appreciate if you can confirm this. Thank you.

**Comment by kostya-sh** [comment]
Thank you for your answer.  Also thank you for correcting my understanding that NAKs are not used on overrun.

Just to clarify further I used StreamingPublisher to make it easy to reproduce and demonstrate the described situation. In our real app this happens much less frequently (~1 time a week with normal conditions and up to once every hour during stress test). We would prefer a slow subscriber to drop old messages in the event of a burst. At the moment however it stops receiving messages for image liveness timeout time (that we set to 1 second). Do you have any recommendation how to configure Aeron for this use case?

**Comment by tmontgomery** [comment]
No worries.

I would suggest that the receiver window should be increased (assuming the default). And probably using the suggested params for throughput performance.

https://github.com/real-logic/aeron/wiki/Performance-Testing (2M window)
https://github.com/real-logic/aeron/tree/master/aeron-samples/scripts

The receiver window defines the overrun. So, a larger window is essential. Especially for bursts. But making it too big will mean issues. So, sizing it appropriately is mandatory for stability.

Beyond that, I would do some performance testing to see what the good param values would be to maintain a very low (or non-existent) back pressure on the Publication under normal situations. This would be a combination of Aeron params and subscriber logic handling. Goal would be to make it so that back pressure only occurs under situations you know it should. i.e. not under bursts unless those bursts are very high (for some definition you probably already have a feel for in the system).

In other words, you only want that situation of a subscriber image going inactive and then active (aka bouncing) when the situation is well beyond the normal and only when it is key for stability.

**Comment by kostya-sh** [comment]
Thank you for the helpful advice.
